### ISSUE TITLE
Suggestion: Re-structure as a group of composable node library packages and remove globals

### ISSUE THREAD

**Issue by fubhy** [NomicFoundation/hardhat#692]
Hi there!

First of all, thanks for working on this. The ethereum developer tooling space is seriously lacking in all regards and I greatly appreciate all the work by everyone here to improve this situation.

I believe that, in order to make this even more useful and to not repeat the same mistakes made by Truffle, Buidler must become group of separate Node library packages instead of striving to be a monolithic command line / runtime. Ideally, I'd like to see a future where I can compose my Solidity development environment exactly for my needs, using the tools and libraries I prefer.

This is made infinitely harder by the fact that Buidler is currently structured, first and foremost, as a rigid command line tool with a lot of imho debatable choices around the runtime environment and the use of magical global properties.

For instance, instead of running `npx buidler test` to wrap and inject this runtime environment into the test runner, a much better approach imho would be to instead offer bridges and plugins for the test runner to instantiate the necessary objects on demand.

This is also highly problematic for Jest for instance (at the time of writing this, Jest adoption is 2x higher than that of Mocha btw.). Jest is a really capable test runner with a lot of great performance improvements and parallelization capabilities to take the best possible advantage of the number of available CPUs, etc. - Wrapping it in an inferior runtime and injecting singleton globals into it the primary process is a failed opportunity.

If Buidler was a group of Node libraries instead of a command line tool, it would be trivial to write a custom test environment, optionally also exposed as a zero-config preset, to integrate with Buidler. While this is somewhat possible already by reaching into the internal modules of Buidler (importing e.g. from `@nomiclabs/buidler/internal/core/providers/construction`) a lot of the assumptions around global configuration and runtime environment are still in the way and impossible to circumvent.

The same is true for all other use-cases: compile, deploy, test, etc. all ultimately suffer from the same upside down architecture that makes extending and composing/customizing this functionality infinitely harder.

Bottom-line: The Node ecosystem offers amazing tools for building powerful, custom command line tools and standalone scripts. Instead of attempting to offer a "do-it-all" runtime I think it should offer the tools to enable people to roll their own scripts and integration layers for existing tools (Jest, Mocha, etc.) instead of wrapping and thereby crippling them.

I'd be happy to assist in planning and executing such an architectural change of direction if this is something you'd consider for your roadmap.

**Comment by adrianmcli** [comment]
I agree 100%. Buidler should take advantage of what the JS ecosystem has to offer instead of trying to blackbox and wrap everything within its own runtime.

As it stands, the major "steps" in Ethereum dapp development is:

- dev (linting, typechecking, etc.)
- compile
- test
- deploy/migrate
- wire up frontend

The JS ecosystem has settled on first-class battle-tested solutions for each of these steps. And if Buidler can become more modular, we should be able to take advantage of the ecosystem. That means instead of running `npx buidler test` or `npx truffle test`, we should be able to run `npm test` which just runs `jest` for example.

I understand Buidler is a task-runner, but so was Gulp and there's a reason why we've moved past it now.

**Comment by kendricktan** [comment]
Agreed. The strengths of being in this ecosystem should be the ability to compose them with one another instead of having each tool create their own abstractions on top of existing tools.

**Comment by fubhy** [comment]
@alcuadrado Hey! I wanted to bring this up again because I still think it would be desirable for hardhat to be a library first and foremost instead. This would require to separate  it from the runtime environment and the magic initialization (or remove that entirely). A lot of super exciting integrations and tools from the broader JS ecosystem are currently not compatible or hard to integrate.

**Comment by roninjin10** [comment]
This would be amazing.  The tradeoff just doesn't make sense to use hardhat's wrapped test runner over vitest or playwright.    Modularity is king.

**Comment by wighawag** [comment]
I agree, personally I see the main "raison d'etre" of hardhat is to be a compilation pipeline + developer oriented eip-1193 provider. Most everything else should be exterior to it.

test should be able to run by any js test runner. this should actually simplify hardhat and encourage more collaboration.

I  got something working with vitest : https://github.com/wighawag/hardhat-vitest-demo but I had to write my own solidity-coverage integration to keep coverage working.

But there are other areas where hardhat could take a less opinionated approach, for example the ethers chai matchers are currently tied strongly to hardhat provider and hardhat-ethers. For some of the functionality it is great but for some it should be independent even of ethers.
